src/main/haskell/language-kore/src/Data/Kore/Proof/Proof.hs

```diff
@@ -0,0 +1,421 @@
+      MembershipForall var a
+          -> ["MembershipForall", pretty var, pretty a]
+      MembershipEq var1 var2
+          -> ["MembershipEq", pretty var1, pretty var2]
+      MembershipNot var a
+          -> ["MembershipNot", pretty var, pretty a]
+      MembershipAnd var a b
+          -> ["MembershipAnd", pretty var, pretty a, pretty b]
+      MembershipExists var1 var2 a
+          -> ["MembershipExists", pretty var1, pretty var2, pretty a]
+      MembershipCong var1 var2 pos a
+          -> ["MembershipCong", pretty var1, pretty var2, pretty pos, pretty a]
+
+instance
+  ( Pretty formula
+  , Pretty (rules subproof)
+  , Pretty subproof
+  , Pretty assumption
+  ) => Pretty (PropF formula rules assumption subproof) where
+    pretty (ByF a b c) = "|- " <> pretty a <> line <> "By " <> pretty b
+
+getConclusion
+    :: Proof
+    -> Term
+getConclusion = conclusion . unFix
+
+getAssumptions
+    :: Proof
+    -> S.Set  Term
+getAssumptions = assumptions . unFix
+
+getFreeVars :: Proof -> S.Set Var
+getFreeVars proof =
+    S.unions
+  $ map freeVars
+  $ S.toList
+  $ getAssumptions proof
+
+assume :: Term -> Proof
+assume formula = By formula (Assumption formula) (S.singleton formula)
+
+-- Given a Rule like `AndIntro a b`, useRule creates a proof, like
+-- By (And a b) (AndIntro a b) (assumptions a `S.union` assumptions b)
+
+-- There are two types of rules:
+-- 1) "special" rules like Assumption, Discharge etc
+-- These are natural deduction introduction/elimination rules which manipulate
+-- quantifiers/assumptions in a tricky way.
+-- 2) "mundane" rules like EqualityElim which merely
+-- generate an axiom using a schema.
+-- The "mundane" rules are passed off to "interpretRule"
+-- which instantiates the schema.
+
+useRule
+  :: Given (MetadataTools Object)
+  => LargeRule Proof
+  -> Proof
+useRule (Assumption formula)
+ = assume formula
+useRule (Discharge hypothesis conclusion)
+ = discharge hypothesis conclusion
+useRule (Abstract var conclusion)
+ | elem var (getFreeVars conclusion) = abstract var conclusion
+ | otherwise = error $ "Variable " ++ show var ++ " appears in assumptions."
+useRule (ExistsElim producer var property (Fix consumer)) =
+    case getConclusion producer of
+      Exists_ _ v p
+       | p == property -> Fix consumer
+            { assumptions = S.delete property $ assumptions consumer
+            }
+       | otherwise -> error "The impossible happened."
+      _ -> impossible
+useRule rule@(OrElim disjunct left right)
+  | getConclusion left == getConclusion right -- FIXME: too picky ==?
+     = let Or_ _ leftAssumption rightAssumption = getConclusion disjunct
+       in By
+          (getConclusion left)
+          rule
+          (S.union
+           (S.delete leftAssumption  $ getAssumptions left)
+           (S.delete rightAssumption $ getAssumptions right)
+          )
+  | otherwise = impossible
+useRule rule =
+  By
+  (interpretRule rule)
```

Comment:
Why no indentation for By's arguments?

Reply:
Sort of looks weird since `By` is a short name.